### PR TITLE
[NPU] Support Qwen3.8-Flash-Next with graph replay, MTP, and Triton kernels

### PR DESCRIPTION
**Pull request by fishAndShrimp**
## Motivation

Enable Qwen3.8-Flash-Next inference on NPU with Qwen Sparse Attention (QSA), graph replay, and NEXTN/MTP speculative decoding, building on the mainline model support from [#37500](https://github.com/sgl-project/sglang/pull/37500).

**Dependency:** [sgl-project/sgl-kernel-npu#807](https://github.com/sgl-project/sgl-kernel-npu/pull/807) provides the Triton kernels used here. Before merging this PR, those kernels must be available in an installation package and the SGLang NPU dependency updated accordingly. The current CI dependency is `2026.9.0`; it does not include these new modules. Local validation uses the commit pair listed below.

## Modifications

- Add NPU QSA reference paths and reuse shared indexer orchestration. Handle NPU KV-cache layouts, fixed graph tensor shapes, live request-to-token mappings, and masked values before attention.
- Dispatch supported sparse attention, paged indexer scoring, block expansion, and hyperconnection operations to `sgl_kernel_npu.qwen3_8_flash_next`. Keep Torch fallback for unsupported input configurations. The new kernel modules must be installed before these NPU paths can run.
- Guard CUDA-only hyperconnection dispatch on NPU, support sigmoid-gated normalization, and disable hyperconnection compilation on NPU to avoid the observed cache-reload failure.
- Make PLE short convolution graph-capturable and handle draft-extend metadata without assuming target-verify fields exist.
- Skip graph input updates only when capture metadata confirms zero recorded operators; propagate background update exceptions to the caller.
- Avoid the CPU-fallback assertion in NPU QSA write planning.
- Add regression coverage for indexing, padding, platform dispatch, fallback, normalization, PLE convolution, graph replay, and graph update error handling. Triton implementations and their standalone tests live in the companion kernel PR.

## Accuracy Tests

acc1

Environment: A3 with TP8, Qwen3.8-Flash-Next BF16, CANN 9.0.0, and torch-npu 2.10.0. Graph replay is enabled; NEXTN uses 3 speculative steps, top-k 1, and 4 draft tokens.

| Implementation | GPQA Diamond status |
| --- | --- |
| Official Qwen3.8-Flash-Next reference ([Qwen model card](https://huggingface.co/Qwen/Qwen3.8-Flash-Next/blob/main/README.md)) | **91.7%**, reported by Qwen. |
| Earlier Torch fallback baseline, SGLang `64b0532544b9f784a5731b1c894d25347acc950b` | 181/198 correct (91.41%); all 198 questions completed. |
| QSA Triton stage (sparse attention, paged MQA, and block expansion), corresponding to SGLang `a02fa50abf08e3ba27d0051ac47b340b37a701a8` | 187/198 correct (94.44%); all 198 questions completed, with zero failed requests. |
| Current Triton integration, SGLang `e2f90ac5f4616cd7d521ef7bc19124dd86ed3b95` + kernel `7ede76da2f3e12fa16f7cad3161b199b24aeb6e1` | 180/198 correct (90.91%); all 198 questions completed, with zero failed requests. |
| Updated NPU integration, SGLang `e58dd411b006b55c62feb37f10c782a64f801ad5` | **91.4%** accuracy over 198 GPQA Diamond questions. |

The official score is an external reference, not a matched evaluation baseline. Our latest result is 0.79 percentage points below the reported 91.7%; equivalence of evaluation settings and repeated-run accuracy has not been established.

The 91.41% result applies to the earlier Torch fallback baseline. Its commit attribution comes from Git history and the run timeline; the evaluation did not embed a Git SHA. It resumed from 197 cached results to complete the remaining question. The current Triton evaluation completed on September 17, 2026, in approximately 8 h 15 min.

The 94.44% run started on September 15 and completed on September 16, 2026. At launch, HEAD was `e95897a8e` with the MQA and block-expansion changes still uncommitted. The saved pre-run development snapshot's file hashes match the corresponding files in the subsequently committed `a02fa50ab`; the attribution is reconstructed from that snapshot and the run timeline, not an embedded evaluation SHA. This result predates HC fusion, the later mainline merge, kernel migration, and the launch-boundary fix. These are individual evaluation runs, not repeated-trial estimates of an accuracy change. The latest result is reported separately above.

GPQA evaluation settings: `temperature=0.9`, `seed=3407`, `n=1`, thinking enabled, maximum 260000 output tokens, and client concurrency 32. The server uses a 4096-token prefill chunk and at most 8 running requests.

Local validation:

| Check | Result and scope |
| --- | --- |
| QSA and hyperconnection integration | 121 tests passed on the current commit pair, including dispatch, fallback, and graph replay. |
| Standalone Triton kernels | 124 tests passed in the companion repository, including large-prefill launch boundaries. |
| Long-KV operator comparison | 4096 query rows with 262144 KV slots passed, including high slot indices, padding, and non-contiguous inputs. This supplemental local check is not yet in the committed suite. |
| Service smoke test | Graph capture and a 6016-token prompt, processed as 4096 + 1920 prefill tokens followed by decode, passed with the same kernel sources as the current commit. |
| Formatting and static checks | Changed files passed pre-commit, including isort, Ruff lint/format, and registered-test validation. Formatting and test-layout fixes are committed. |

The completed GPQA run used the new Triton sources with the previously installed native library. Separate source-build validation on a second machine is reported below. The operator checks alone do not establish full-model 256k context or complete request eviction/resumption coverage.

### Native build validation

On a second A3/aarch64 machine, the native kernel library and wheel were successfully built from kernel commit `7ede76da2f3e12fa16f7cad3161b199b24aeb6e1`, using CANN 9.0.0, Python 3.11.15, PyTorch/torch-npu 2.10.0, GCC 11.4.0, and CMake 3.22.1. The optional CATLASS component was disabled.

- The wheel contains the newly built native library and all four Triton modules. Its library matches the source-tree build output byte for byte.
- The wheel was **not installed**. Validation loaded the source tree and its newly built native library; the global installation was left unchanged.
- Native FP16 addition checks passed exactly for lengths 2048 and 8192. All 124 standalone Triton tests and 121 SGLang integration tests passed again on this machine.
- All eight TP ranks were verified to load the new library. Service health, a short request, and a 4216-token prompt processed as 4096 + 120 prefill tokens followed by decode passed.

No tracked source changes were required for the build.

## Speed Tests and Profiling

prof1

### Serving benchmark on the second machine

The serving benchmark used TP8 with graph replay and NEXTN/MTP. It completed all **80 requests** using `sglang-oai-chat`, random inputs/outputs with configured lengths of 1024/1024 and range ratio 0.5, maximum concurrency 8, unlimited offered request rate, and one warmup request. Request-body temperature was explicitly set to **0.01**. Actual token totals are shown below because the randomized lengths are not fixed at 1024.

| Metric | Result |
| --- | ---: |
| Successful requests | 80/80 |
| Duration | 945.87 s |
| Input tokens | 62,912 |
| Generated output tokens | 62,023 |
| Aggregate output throughput | **65.57 tokens/s** |
| Aggregate input + output throughput | 132.08 tokens/s |
| Mean / median TTFT | 6241.56 / 3821.73 ms |
| Mean TPOT | 112.24 ms |
| Mean end-to-end request latency | 92.63 s |
| Mean speculative acceptance length | 2.65 |

This is an absolute serving measurement, not a before/after speedup. The second-machine checkout/build record identifies the commit pair above; the benchmark JSON records installed package version metadata rather than a runtime source Git SHA. This benchmark's workload and temperature differ from GPQA and the earlier development comparisons.

### Earlier incremental measurements

Local development measurements used TP8, graph replay, NEXTN/MTP, eight concurrent requests, 512 output tokens per request, and non-greedy sampling (`temperature=0.9`, `seed=3407`). Each comparison used the same prompts and two rounds; throughput is total output tokens divided by total elapsed time.

| Incremental change | Before (tokens/s) | After (tokens/s) | Improvement |
| --- | ---: | ---: | ---: |
| Paged MQA Triton scoring, with sparse-attention Triton already enabled | 49.31 | 66.34 | 34.53% |
| HC fusion, with QSA Triton paths already enabled | 65.64 | 68.76 | 4.76% |

These are separate incremental comparisons made before kernel migration. Their gains should not be added together or interpreted as a controlled full-model Torch-versus-Triton speedup. Non-greedy speculative acceptance and batch evolution can affect these short measurements.

After migration and the large-prefill launch fix, a smoke run on the current sources completed eight requests of 512 output tokens at **67.26 aggregate output tokens/s**. This is a regression observation, not a controlled speedup claim or full GPQA throughput.

Both sides of each earlier comparison retain the same non-greedy sampling behavior. The completed GPQA result and the second-machine serving benchmark are reported separately above.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35719706012](https://github.com/sgl-project/sglang/actions/runs/35719706012)
Latest PR Test (Extra): :x: [Run #35719705401](https://github.com/sgl-project/sglang/actions/runs/35719705401)
Latest PR Test (AMD ROCm 10): :x: [Run #35719705900](https://github.com/sgl-project/sglang/actions/runs/35719705900)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/graph_runner/npu_cudagraph_backend.py [hunk 1]
```diff
@@ -150,30 +150,75 @@ def replay_with_input_update(
...
+        except (AttributeError, TypeError) as e:
+            raise RuntimeError(
+                "Cannot inspect NPU graph update records; "
+                "check torch_npu graph API compatibility"
+            ) from e
+        if not auto_dispatch_capture:
+            raise RuntimeError("NPU graph updates require auto_dispatch_capture=True")
+        # With no update records, skip graph.update() and replay directly.
+        if update_count == 0:
+            graph.replay()
```
Comment (fishAndShrimp) [inline python/sglang/srt/hardware_backend/npu/graph_runner/npu_cudagraph_backend.py]: keep 1 method

File: python/sglang/srt/layers/attention/qsa/kernel.py [hunk 2]
```diff
@@ -50,13 +59,33 @@ def qsa_fast_topk(
...
         )
 
-    # CPU/reference path mirrors the CUDA operator's fixed-width, relative output.
     output = torch.full(
         (logits.shape[0], topk),
         -1,
         dtype=torch.int32,
         device=logits.device,
     )
+    if _is_npu:
```
Comment (fishAndShrimp) [inline python/sglang/srt/layers/attention/qsa/kernel.py]: check triton impl

File: python/sglang/srt/layers/attention/qsa/kernel.py [hunk 3]
```diff
@@ -249,6 +264,18 @@ def expand_qsa_block_indices(
     rows = block_indices.shape[0]
     if query_positions.numel() != rows or sequence_lengths.numel() != rows:
         raise ValueError("query positions and sequence lengths must match top-k rows")
+    if _is_npu:
+        from sgl_kernel_npu.qwen3_8_flash_next.qsa_expansion import expand_blocks
+
+        # The package wrapper owns the model contract and its internal dispatch.
```
Comment (fishAndShrimp) [inline python/sglang/srt/layers/attention/qsa/kernel.py]: rename to triton_expand_qsa_block_indices

File: python/sglang/srt/layers/attention/qsa/qsa_indexer.py [hunk 4]
```diff
@@ -628,6 +632,18 @@ def forward_cuda(
...
 
+    def forward_npu(
+        self,
+        hidden_states: torch.Tensor,
+        positions: torch.Tensor,
+        forward_batch,
+        indexer_metadata,
+    ) -> torch.Tensor:
+        # Share the orchestration; individual operators select Torch on NPU.
+        return self.forward_cuda(
```
Comment (fishAndShrimp) [inline python/sglang/srt/layers/attention/qsa/qsa_indexer.py]: verify the reusing of forward_cuda

File: python/sglang/srt/layers/hyperconnection.py [hunk 5]
```diff
@@ -216,8 +226,25 @@ def _combine_compute(
...
+        # lookup is redirected to NPU by transfer_to_npu. A process group without
+        # an NPU backend then raises "No backend type associated with device
+        # type npu". Thus fresh compilation succeeds, but a cached restart fails
+        # in the extra timeout step, not in the compiled mix/combine operations.
+        # Keep these functions outside torch.compile on NPU; supported vector
+        # operations can still dispatch to NPU Triton kernels.
+        # Other platforms keep compilation; server NPU graph capture is unchanged.
+        # TODO: Remove disable=_is_npu once dependency compatibility is verified
+        # by both fresh-cache and cached-restart tests on NPU. Newer torch_npu
+        # source has related timeout handling, but an upgrade alone is not proof.
```
Comment (fishAndShrimp) [inline python/sglang/srt/layers/hyperconnection.py]: simlify comment

File: python/sglang/srt/layers/attention/qwen_sparse_attn_backend.py [hunk 6]
```diff
@@ -37,13 +37,49 @@
...
+from sglang.srt.utils import is_npu
 
 logger = logging.getLogger(__name__)
 
+_is_npu = is_npu()
 
 _TRTLLM_SPARSE_PAGE_SIZE = 64
 
 
+def _flatten_qsa_kv_cache(cache: torch.Tensor, name: str) -> torch.Tensor:
```
Comment (fishAndShrimp) [inline python/sglang/srt/layers/attention/qwen_sparse_attn_backend.py]: ndim=3 condition?
